etcd-backup/main.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/spf13/cobra"
+	"go.etcd.io/etcd/client/pkg/v3/transport"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/snapshot"
+
+	"go.uber.org/zap"
+)
+
+const (
+	DefaultDialTimeout       = 15 * time.Second
+	DefaultEtcdClientTimeout = 5 * time.Minute
+)
+
+type options struct {
+	clusterID      string
+	etcdEndpoint   string
+	s3BucketName   string
+	s3BucketRegion string
+}
+
+func main() {
+	var opts options
+
+	cmd := &cobra.Command{
+		Use:          "etcd-backup",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return run(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.clusterID, "cluster-id", "", "ID of the hosted cluster, used as a key for the snapshot file in S3")
+	cmd.Flags().StringVar(&opts.etcdEndpoint, "etcd-endpoint", "", "endpoint of the etcd cluster to backup.")
+	cmd.Flags().StringVar(&opts.s3BucketName, "s3-bucket-name", "", "name of the S3 bucket to store etcd backups.")
+	cmd.Flags().StringVar(&opts.s3BucketRegion, "s3-bucket-region", "", "AWS region of the S3 bucket to store etcd backups.")
+
+	cmd.MarkFlagRequired("cluster-id")
+	cmd.MarkFlagRequired("etcd-endpoint")
+	cmd.MarkFlagRequired("s3-bucket-name")
+
+	if err := cmd.Execute(); err != nil {
+		os.Exit(1)
+	}
+}
+
+// go run etcd-backup/main.go etcd-backup --cluster-id mraee-hosted-cluster --etcd-endpoint localhost:2379 --s3-bucket-name mraee-hypershift-bucket
+func run(opts options) error {
+	tlsInfo := transport.TLSInfo{
+		CertFile:      "/etc/etcd/tls/client/etcd-client.crt",
+		KeyFile:       "/etc/etcd/tls/client/etcd-client.key",
+		TrustedCAFile: "/etc/etcd/tls/etcd-ca/ca.crt",
+	}
+
+	tlsConfig, err := tlsInfo.ClientConfig()
+	if err != nil {
+		return fmt.Errorf("error creating client TLSConfig: %w", err)
+	}
+
+	cfg := clientv3.Config{
+		Endpoints:   []string{opts.etcdEndpoint},
+		DialTimeout: DefaultDialTimeout,
+		TLS:         tlsConfig,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), DefaultEtcdClientTimeout)
+	defer cancel()
+
+	filePath := "/tmp/snapshot.db"
+	err = snapshot.Save(ctx, zap.L(), cfg, filePath)
+	if err != nil {
+		return fmt.Errorf("failed to snapshot etcd: %w", err)
+	}
+
+	return uploadToS3(opts, filePath)
+}
+
+func uploadToS3(opts options, snapshotFilePath string) error {
+	config := aws.NewConfig()
+	// AWS_REGION must be set if s3BucketRegion is empty
+	if opts.s3BucketRegion != "" {
+		config.Region = aws.String(opts.s3BucketRegion)
+	}
+	awsSession := session.Must(session.NewSession(config))
+
+	f, err := os.Open(snapshotFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to open file %q, %v", snapshotFilePath, err)
+	}
+
+	key := fmt.Sprintf("%s_%d.db", opts.clusterID, time.Now().Unix())
```

Comment:
I would think we'd want something like `/[cluster-id]/[ts].db`

Reply:
what is cluster id here, I assume ObjectMeta UID which is the only value truly guaranteed to be unique, assuming the aws account is scoped to a single management cluster.
Can we agree and document this?

Reply:
I was thinking either `hostecCluster.spec.ClusterID` ,the ObjectMeta UID or ocm cluster id (don't know if we can get this one)

Reply:
Furthermore there will have to be an additional prefix to designate the backup period i.e. `hourly`, `daily`, `weekly`. The broader logic is not here yet to accommodate multiple backup periods, but we will need some prefixing strategy to apply corresponding lifecycle rules to the backup objects. Prefix name would probably have to be configurable if the backup schedule is configurable.